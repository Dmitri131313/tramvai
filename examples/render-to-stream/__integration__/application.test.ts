import type { StartCliResult } from '@tramvai/test-integration';
import { startCli } from '@tramvai/test-integration';
import { initPlaywright } from '@tramvai/test-pw';
import path from 'path';
import { Writable } from 'stream';

jest.setTimeout(30000);

class WritableBuffer extends Writable {
  private chunks: string[] = [];

  getChunks() {
    return this.chunks;
  }

  _write(chunk, encoding, callback) {
    this.chunks.push(chunk.toString('utf-8'));
    callback();
  }
}

describe('render-to-stream', () => {
  let app: StartCliResult;

  beforeAll(async () => {
    app = await startCli('render-to-stream', {
      rootDir: path.resolve(__dirname, '../'),
    });
  }, 80000);

  afterAll(() => {
    return app.close();
  });

  describe('streaming response', () => {
    // eslint-disable-next-line jest/no-done-callback
    it('deferred page', (done) => {
      const stream = new WritableBuffer();

      const initialChunksContent = [
        '<!DOCTYPE html>',
        // actions, resolved before response
        `window.__TRAMVAI_DEFERRED_ACTIONS['fastDeferred'].resolve({"data":"ok"});</script>`,
        `window.__TRAMVAI_DEFERRED_ACTIONS['failedFastDeferred'].reject({"executionContextName":"root.command-line:resolve_page_deps.pageActions.failedFastDeferred","message":"Failed Fast Deferred"`,
        '<div class="application">',
        // lazy component outside Suspense
        'Deferred Page',
        `<script>typeof window.__TRAMVAI_DEFERRED_HYDRATION === 'function' ? window.__TRAMVAI_DEFERRED_HYDRATION() : window.__TRAMVAI_DEFERRED_HYDRATION = true;</script>`,
      ];
      const deferredChunksContent = [
        // deferred actions promises teleportation
        `<script>window.__TRAMVAI_DEFERRED_ACTIONS['longDeferred'].resolve({"data":"ok"});</script>`,
        `<script>window.__TRAMVAI_DEFERRED_ACTIONS['failedDeferred'].reject({"executionContextName":"root.command-line:resolve_page_deps.pageActions.failedDeferred","message":"Failed Deferred"`,
        // lazy component inside Suspense, blocking assets
        'components-features-Data-Data.chunk.css',
        'components-features-Data-Data.chunk.js',
        `Response: ok`,
        '</body></html>',
      ];

      setTimeout(() => {
        const initialHTML = stream.getChunks().join('\n');

        // all critical scripts async for streaming
        expect(initialHTML.match(/data-critical="true" async="async"/g)?.length).toBe(5);

        initialChunksContent.forEach((chunk) => {
          expect(initialHTML).toContain(chunk);
        });

        deferredChunksContent.forEach((chunk) => {
          expect(initialHTML).not.toContain(chunk);
        });
      }, 500);

      app
        .request('/deferred/')
        .expect(200)
        .expect('Transfer-Encoding', 'chunked')
        .pipe(stream)
        .on('finish', (error) => {
          if (error) {
            return done(error);
          }

          const finalHTML = stream.getChunks().join('\n');

          deferredChunksContent.forEach((chunk) => {
            expect(finalHTML).toContain(chunk);
          });

          done();
        });
    });
  });

  describe('server-side rendering', () => {
    it('main page', async () => {
      const { parsed } = await app.render('/');

      expect(parsed.innerHTML).toMatchSnapshot();
    });

    it('deferred page', async () => {
      const { parsed } = await app.render('/deferred/');

      expect(parsed.innerHTML).toMatchSnapshot();
    });

    it('non-deferred page', async () => {
      const { parsed } = await app.render('/non-deferred/');

      expect(parsed.innerHTML).toMatchSnapshot();
    });
  });

  describe('client hydration', () => {
    it('main page', async () => {
      const { browser, getPageWrapper } = await initPlaywright(app.serverUrl);

      const { page } = await getPageWrapper(app.serverUrl);

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
              "Tramvai ðŸ¥³
              Main
              Second
              Deferred
              Non-deferred
              Main Page
              Child Component
              Error boundary
              this Footer in render-to-stream

              This is modal for index page!"
          `);

      await browser.close();
    });

    it('deferred page', async () => {
      const { browser, getPageWrapper } = await initPlaywright(app.serverUrl);

      const { page } = await getPageWrapper(`${app.serverUrl}/deferred/`);

      expect(await page.$eval('.application', (node) => (node as HTMLElement).innerText))
        .toMatchInlineSnapshot(`
        "Tramvai ðŸ¥³
        Main
        Second
        Deferred
        Non-deferred
        Deferred Page
        Response: ok
        Response: ok
        Error: Failed Fast Deferred
        Error: Failed Deferred
        this Footer in render-to-stream"
      `);

      await browser.close();
    });
  });
});
