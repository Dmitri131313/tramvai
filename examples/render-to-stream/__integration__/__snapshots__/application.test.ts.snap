// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render-to-stream server-side rendering deferred page 1`] = `
"<!DOCTYPE html>
<html class="no-js" lang="ru">
  <head>
    <meta charset="UTF-8" >
    
    <title data-meta-dynamic="true">Tramvai</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    

    

    

    
    <link crossorigin="anonymous"
      data-critical="true"
      rel="stylesheet"
      href="\${STATIC_URL}/dist/client/components-features-Title-Title.chunk.css"
    >
    <link crossorigin="anonymous" data-critical="true" rel="stylesheet" href="\${STATIC_URL}/dist/client/platform.css" >
    

    

    

    
    <script id="__TRAMVAI_STATE__" type="application/json"></script>
    

    
    <script></script>
    <script></script>
    <script></script>
    <script></script>
    <script></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/@_routes_deferred_index.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/components-features-Title-Title.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/hmr.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  </head>
  <body>
    

    

    
    <div class="application">
      <div>
        <h1>
          Tramvai
          <span role="img" aria-label="Salute">ðŸ¥³</span>
        </h1>
        <ul>
          <li><a href="/">Main</a></li>
          <li><a href="/second/">Second</a></li>
          <li><a href="/deferred/">Deferred</a></li>
          <li><a href="/non-deferred/">Non-deferred</a></li>
          <li><a href="/deferred/foo/">Deferred Foo</a></li>
          <li><a href="/deferred/bar/">Deferred Bar</a></li>
          <li><a href="/deferred/baz/">Deferred Baz</a></li>
        </ul>
      </div>
      <h3 class="Title__title_jjG8q">Deferred Page</h3>
      <template id="B:0"></template>
      <div>Loading long...</div>
      
      <div>
        Response:
        ok
      </div>
      
      <div>
        Error:
        Failed Fast Deferred
      </div>
      <template id="B:1"></template>
      <div>Loading long error...</div>
      
      <div class="Footer__footer_L7U1b"><div>this Footer in render-to-stream</div></div>
      <script></script>
      <script></script>
      <script></script>
      <link rel="stylesheet"
        href="\${STATIC_URL}/dist/client/components-features-Data-Data.chunk.css"
        crossorigin="anonymous"
        data-critical="true"
      >
      <script src="\${STATIC_URL}/dist/client/components-features-Data-Data.chunk.js"
        charset="utf-8"
        crossorigin="anonymous"
        data-critical="true"
      ></script>
      <div hidden id="S:0"><div class="Data__data_Q_cJh">Response: ok</div></div>
      <script></script>
      <div hidden id="S:1">
        <div>
          Error:
          Failed Deferred
        </div>
      </div>
      <script></script>
    </div>
    

    

    

    

    

    

    
  </body>
</html>
"
`;

exports[`render-to-stream server-side rendering main page 1`] = `
"<!DOCTYPE html>
<html class="no-js" lang="ru">
  <head>
    <meta charset="UTF-8" >
    
    <title data-meta-dynamic="true">Tramvai</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    

    

    

    
    <link crossorigin="anonymous"
      data-critical="true"
      rel="stylesheet"
      href="\${STATIC_URL}/dist/client/@_routes_index.chunk.css"
    >
    <link crossorigin="anonymous" data-critical="true" rel="stylesheet" href="\${STATIC_URL}/dist/client/platform.css" >
    

    

    

    
    <script id="__TRAMVAI_STATE__" type="application/json"></script>
    

    
    <script></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/@_routes_index.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/hmr.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  </head>
  <body>
    

    

    
    <div class="application">
      <div>
        <h1>
          Tramvai
          <span role="img" aria-label="Salute">ðŸ¥³</span>
        </h1>
        <ul>
          <li><a href="/">Main</a></li>
          <li><a href="/second/">Second</a></li>
          <li><a href="/deferred/">Deferred</a></li>
          <li><a href="/non-deferred/">Non-deferred</a></li>
          <li><a href="/deferred/foo/">Deferred Foo</a></li>
          <li><a href="/deferred/bar/">Deferred Bar</a></li>
          <li><a href="/deferred/baz/">Deferred Baz</a></li>
        </ul>
      </div>
      <div>
        Main Page
        <div>Child Component</div>
        <template data-msg="Error"
          data-stck="
    at ErrorComponent (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at Suspense
    at ErrorBoundary (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at div
    at MainPage
    at InnerLoadable (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at LoadableWithChunkExtractor
    at Loadable
    at UniversalErrorBoundary (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at PageErrorBoundary (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at PageRenderComponent (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at RenderChildrenComponent (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at NestedLayoutRenderComponent (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at RenderChildren (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at RenderChildren (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at RenderChildren (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at Layout (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at LayoutRenderComponent (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at Root (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at Provider (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at Provider (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])
    at ChunkExtractorManager (\${STATIC_URL}/dist/server/server.js:[LINE]:[SYMBOL])"
        ></template>Error fallback
      </div>
      <div class="Footer__footer_L7U1b">
        <div>this Footer in render-to-stream</div>
        <p class="main__MainModal_wPdhA">This is modal for index page!</p>
      </div>
      <script></script>
    </div>
    

    

    

    

    

    

    
  </body>
</html>
"
`;

exports[`render-to-stream server-side rendering non-deferred page 1`] = `
"<!DOCTYPE html>
<html class="no-js" lang="ru">
  <head>
    <meta charset="UTF-8" >
    
    <title data-meta-dynamic="true">Tramvai</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" >
    

    

    

    
    <link crossorigin="anonymous" data-critical="true" rel="stylesheet" href="\${STATIC_URL}/dist/client/platform.css" >
    

    

    

    
    <script id="__TRAMVAI_STATE__" type="application/json"></script>
    

    
    <script></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/@_routes_non-deferred_index.chunk.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/hmr.js"
    ></script>
    <script charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      async="async"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  </head>
  <body>
    

    

    
    <div class="application">
      <div>
        <h1>
          Tramvai
          <span role="img" aria-label="Salute">ðŸ¥³</span>
        </h1>
        <ul>
          <li><a href="/">Main</a></li>
          <li><a href="/second/">Second</a></li>
          <li><a href="/deferred/">Deferred</a></li>
          <li><a href="/non-deferred/">Non-deferred</a></li>
          <li><a href="/deferred/foo/">Deferred Foo</a></li>
          <li><a href="/deferred/bar/">Deferred Bar</a></li>
          <li><a href="/deferred/baz/">Deferred Baz</a></li>
        </ul>
      </div>
      <div>Loading...</div>
      <div class="Footer__footer_L7U1b"><div>this Footer in render-to-stream</div></div>
      <script></script>
    </div>
    

    

    

    

    

    

    
  </body>
</html>
"
`;
